pkg/ipcache/ipcache.go

```diff
@@ -457,24 +458,61 @@ func (ipc *IPCache) DumpToListener(listener IPIdentityMappingListener) {
 	ipc.RUnlock()
 }
 
+// MU is a batched metadata update, the short name is to cut down on visual clutter.
+type MU struct {
+	Prefix   netip.Prefix
+	Source   source.Source
+	Resource ipcachetypes.ResourceID
+	Metadata []IPMetadata
+}
+
 // UpsertMetadata upserts a given IP and some corresponding information into
 // the ipcache metadata map. See IPMetadata for a list of types that are valid
 // to pass into this function. This will trigger asynchronous calculation of
 // any datapath updates necessary to implement the logic associated with the
 // specified metadata.
 func (ipc *IPCache) UpsertMetadata(prefix netip.Prefix, src source.Source, resource ipcacheTypes.ResourceID, aux ...IPMetadata) {
+	ipc.UpsertMetadataBatch(MU{Prefix: prefix, Source: src, Resource: resource, Metadata: aux})
+}
+
+// UpsertMetadataBatch applies updates to multiple prefixes in a single transaction,
+// reducing potential lock contention.
+func (ipc *IPCache) UpsertMetadataBatch(updates ...MU) {
+	prefixes := make([]netip.Prefix, 0, len(updates))
 	ipc.metadata.Lock()
-	ipc.metadata.upsertLocked(prefix, src, resource, aux...)
+	for _, upd := range updates {
+		ipc.metadata.upsertLocked(upd.Prefix, upd.Source, upd.Resource, upd.Metadata...)
+		prefixes = append(prefixes, upd.Prefix)
+	}
 	ipc.metadata.Unlock()
-	ipc.metadata.enqueuePrefixUpdates(prefix)
+	ipc.metadata.enqueuePrefixUpdates(prefixes...)
 	ipc.TriggerLabelInjection()
 }
 
+// RemoveMetadata removes metadata associated with a specific resource from the
+// supplied prefix. Individual metadata types must be supplied for removal, but the
+// data need not match.
```

Comment:
I guess the data need not match, but the data *type* must match because it hits `resourceInfo.unmerge()`.

Reply:
Yup, I tried to capture this in the example.